### PR TITLE
[ffigen] [jnigen] Util to dump entire parsed API

### LINKED ISSUE #3605: [ffigen][jnigen] Util to dump entire parsed API
We tell people to write a Dart config script for FFIgen/JNIgen, and give them a bunch of examples. But if you write a config, and the API elements you expect don't appear in your bindings, it's unclear what to do next.

One debugging approach, under the new visitor API, is to write a custom visitor that just prints all the API elements. Then you can see if the problem is that you haven't set the correct entry points (headers etc), or your filters are misconfigured.

We should provide a util that does this, so users don't have to do it. Best API is probably an extension method on `FfiGenerator`/`JniGenerator` that dumps all the AST nodes (public AST nodes, as seen by a custom visitor). Eg:

```
Func(originalName, USR)
ObjCInterface(originalName, USR)
ObjCMethod(ObjCInterface(originalName, USR), methodSelector)
...
```

That output is not the most human readable format, but it's clear and greppable. Other suggestions welcome.

### PR DESCRIPTION
**Pull request by amanmaurya92**
## Description

When configuring FFIgen or JNIgen, users frequently encounter scenarios where expected API elements do not appear in the generated bindings (due to incorrect entry point headers, missing include paths, preprocessor flags, or misconfigured filters). Diagnosing this previously required users to construct custom AST visitor debug scripts from scratch.

This PR introduces a standardized AST dumping utility:
- Adds `ApiDumperVisitor` to `package:ffigen` to traverse and format all public AST declarations into a clear, greppable format (`Node(name, USR/parent)`).
- Adds `dumpApi({StringSink? sink, Logger? logger, Uri? libclangDylib})` extension method on `FfiGenerator` that parses headers without writing output binding files.
- Adds matching `ApiDumperVisitor` and `dumpApi({StringSink? sink, Logger? logger})` on `JniGenerator` in `package:jnigen` to dump Java classes, methods, parameters, and fields.
- Adds comprehensive unit tests in both `pkgs/ffigen` and `pkgs/jnigen` covering synthetic AST node dumping and end-to-end header parsing.
- Updates `CHANGELOG.md` for both packages.

## Related Issues

Fixes https://github.com/dart-lang/native/issues/3605

## PR Checklist

- [x] I’ve reviewed the [contributor guide](https://github.com/dart-lang/native/blob/main/CONTRIBUTING.md) and applied the relevant portions to this PR.
- [x] I've run `dart tool/ci.dart --all` locally and resolved all issues identified. This ensures the PR is formatted, has no lint errors, and ran all code generators. This applies to the packages part of the toplevel `pubspec.yaml` workspace.
- [x] All existing and new tests are passing. I added new tests to check the change I am making.
- [x] The PR is actually solving the issue. PRs that don't solve the issue will be closed. Please be respectful of the maintainers' time. If it's not clear what the issue is, feel free to ask questions on the GitHub issue before submitting a PR.
- [x] I have updated `CHANGELOG.md` for the relevant packages. (Not needed for small changes such as doc typos).
- [ ] I have [updated the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change) if necessary.

### CODE REVIEW

File: pkgs/ffigen/lib/src/config_provider/api_dumper.dart [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  static String _declString(DeclNode decl) {
+    if (decl is ObjCCategory) {
+      final interfaceStr = _declString(decl.interface);
+      return 'ObjCCategory($interfaceStr, ${decl.originalName}, ${decl.usr})';
+    }
+    return '${decl.runtimeType}(${decl.originalName}, ${decl.usr})';
+  }
+
+  static String _namedNodeString(NamedNode node) {
+    if (node is DeclNode) {
```
Comment (liamappelbe) [inline pkgs/ffigen/lib/src/config_provider/api_dumper.dart]: Might be cleaner to use a switch here.

File: pkgs/ffigen/lib/src/config_provider/api_dumper.dart [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    final buffer = StringBuffer();
+    final visitor = ApiDumperVisitor(buffer);
+    visitor.visitAll(nodes);
+    return buffer.toString();
+  }
+
+  static String _declString(DeclNode decl) {
+    if (decl is ObjCCategory) {
+      final interfaceStr = _declString(decl.interface);
+      return 'ObjCCategory($interfaceStr, ${decl.originalName}, ${decl.usr})';
```
Comment (liamappelbe) [inline pkgs/ffigen/lib/src/config_provider/api_dumper.dart]: Looking at the outputs in the test, I think it'd be more readable to put the `$interfaceStr` last. Same with the parents below.

File: pkgs/ffigen/lib/src/config_provider/api_dumper.dart [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  /// Dumps all [nodes] and returns the formatted output string.
+  static String dump(Iterable<AstNode> nodes) {
+    final buffer = StringBuffer();
+    final visitor = ApiDumperVisitor(buffer);
+    visitor.visitAll(nodes);
+    return buffer.toString();
+  }
+
+  static String _declString(DeclNode decl) {
+    if (decl is ObjCCategory) {
```
Comment (liamappelbe) [inline pkgs/ffigen/lib/src/config_provider/api_dumper.dart]: Looks like this special casing exists here because `ObjCMethod.parent` can be either a `ObjCInterface` or `ObjCCategory`. So it would be better to do that special casing in `visitObjCMethod`:

```dart
@override
void visitObjCMethod(ObjCMethod node) {
  _sink.writeln('ObjCMethod(${_objCMethodParent(node.parent)}, ${node.selector})');
}

static String _objCMethodParent(DeclNode node) => switch(node) {
  ObjCInterface i => _visitObjCInterface(i),
  ObjCCategory c => _visitObjCCategory(c),
};

@override
void visitObjCInterface(ObjCInterface node) {
  _sink.writeln(_visitObjCInterface(node));
}

static String _visitObjCInterface(ObjCInterface node) => _declString(node);

static String _declString(DeclNode decl) =>
  '${decl.runtimeType}(${decl.originalName}, ${decl.usr})';
```

Same for `_namedNodeString`.

File: pkgs/ffigen/lib/src/config_provider/config.dart [hunk 4]
```diff
@@ -133,6 +133,20 @@ final class FfiGenerator {
       this,
     ).generate(logger: logger, libclangDylib: libclangDylib);
   }
+
+  /// Parses the headers according to this config and dumps all user-visible
+  /// AST nodes to [sink] (or stdout if not provided).
+  ///
+  /// Returns the formatted dump string.
```
Comment (liamappelbe) [inline pkgs/ffigen/lib/src/config_provider/config.dart:140]: I think the API should do one or the other: return the string or print to the sink. We don't need both.

File: pkgs/ffigen/lib/src/ffigen.dart [hunk 5]
```diff
@@ -8,13 +8,48 @@ import 'package:cli_util/cli_logging.dart' show Ansi;
...
+    logger ??= createDefaultLogger();
+    final buffer = StringBuffer();
+    final dumper = ApiDumperVisitor(buffer);
+
+    final dumperConfig = FfiGenerator(
+      input: input,
+      cpp: cpp,
+      objectiveC: objectiveC,
+      output: output,
+      visitors: [...visitors, dumper],
```
Comment (liamappelbe) [inline pkgs/ffigen/lib/src/ffigen.dart]: Why run the user visitors here?

File: pkgs/ffigen/CHANGELOG.md [hunk 6]
```diff
@@ -38,6 +38,7 @@
   ObjC category methods returning `instancetype` are filtered.
 - Minor Objective-C code generator and function type signature fixes.
 - Bump `package:code_assets` dependency to `^2.0.0`.
+- Add `dumpApi` utility and `ApiDumperVisitor` to dump entire parsed AST.
```
Comment (liamappelbe) [inline pkgs/ffigen/CHANGELOG.md]: v22 is already out. You'll need to bump the version to `22.1.0-wip`, here and in the pubspec.

File: pkgs/jnigen/lib/src/elements/api_dumper.dart [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    final parentName = _currentClass?.binaryName ?? 'unknown';
+    _sink.writeln('Field(ClassDecl($parentName), ${field.originalName})');
+  }
+
+  @override
+  void visitParam(Param parameter) {
+    final parentClass = _currentClass?.binaryName ?? 'unknown';
+    final parentMethod = _currentMethod?.originalName ?? 'unknown';
+    _sink.writeln(
+      'Param(Method(ClassDecl($parentClass), $parentMethod), '
```
Comment (liamappelbe) [inline pkgs/jnigen/lib/src/elements/api_dumper.dart]: DRY. Reuse the logic from `visitMethod`.

File: pkgs/jnigen/lib/src/elements/api_dumper.dart [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  void visitClass(ClassDecl c) {
+    _currentClass = c;
+    _sink.writeln('ClassDecl(${c.binaryName})');
+  }
+
+  @override
+  void visitMethod(Method method) {
+    _currentMethod = method;
+    final parentName = _currentClass?.binaryName ?? 'unknown';
+    _sink.writeln('Method(ClassDecl($parentName), ${method.originalName})');
```
Comment (liamappelbe) [inline pkgs/jnigen/lib/src/elements/api_dumper.dart]: DRY. Reuse the logic from `visitClass`.

Comment (liamappelbe) [inline pkgs/jnigen/lib/src/elements/api_dumper.dart]: As in the FFIgen one, move the method name to the start.

File: pkgs/ffigen/lib/src/config_provider/config.dart (hunk 4 above)
Comment (amanmaurya92) [reply pkgs/ffigen/lib/src/config_provider/config.dart:140]: The dual behavior was for convenience, but doing one or the other makes complete sense for a cleaner API. so between:
`Future<String> dumpApi(...)` and 
`Future<void> dumpApi({StringSink? sink})`
Which do you prefer? I was leaning more towards Future<String> dumpApi(...).

Comment (liamappelbe) [reply pkgs/ffigen/lib/src/config_provider/config.dart:140]: Yeah, I think returning a `String` makes more sense.

Comment (amanmaurya92) [reply pkgs/ffigen/lib/src/config_provider/config.dart:140]: I have addressed that now. thanks

File: pkgs/ffigen/lib/src/ffigen.dart (hunk 5 above)
Comment (amanmaurya92) [reply pkgs/ffigen/lib/src/ffigen.dart]: My idea was that if a user has configured renaming or filtering visitors, running them before `dumper` allows `dumpApi()` to inspect the AST in the exact state when the code generator will see it (consistent with how jnigen runs visitors before dumper in generate_bindings.dart). if you prefer I can remove it or keep `visitors: [dumper]`.

Comment (liamappelbe) [reply pkgs/ffigen/lib/src/ffigen.dart]: I can see the utility in that, but I don't think it will work. At least not the way you have it written currently.

In FFIgen, later visitors still see nodes that have been filtered out by earlier visitors, so this won't actually show which nodes are filtered. You could fix this by only printing the node if `isIncluded`, or printing `"X "` or something at the start of the line if the node isn't included. This is an inconsistency with JNIgen, so maybe we should change the visitor infra to skip nodes that are filtered, but it's tricky because we filter almost all nodes out by default, so the first visitor wouldn't see anything :upside_down_face:.

As for renames, you're printing the `originalName`, not the `name`, so these changes won't be visible either. As it stands now I don't think any of the user visitors can have any visible effect on the `dumpApi` output.

In any case, if this is a use case you want to support, you'll need to write tests for it.

Comment (amanmaurya92) [reply pkgs/ffigen/lib/src/ffigen.dart]: Ohh okay, 😅 must have missed that. I will look into it more

Comment (amanmaurya92) [reply pkgs/ffigen/lib/src/ffigen.dart]: and you're completely right. I verified the parser pipeline. user visitors don't have any visible effect on the output anyway. I've updated `dumpApi` to only run `dumper` in both ffigen and jnigen.

**Comment by amanmaurya92** [comment]
@liamappelbe all CI are green now, please merge this, as this is my first pr here 😄. I appreciate your guidance.

File: pkgs/jnigen/lib/src/elements/j_elements.dart [hunk 9]
```diff
@@ -153,6 +153,9 @@ class Method implements _Element {
   /// Whether this method is a constructor.
   bool get isConstructor => _method.isConstructor;
 
+  /// The method descriptor (e.g. `(Ljava/lang/String;)V`), if available.
```
Comment (liamappelbe) [inline pkgs/jnigen/lib/src/elements/j_elements.dart:156]: This change seems unrelated?

If you want to include it, it'll need a changelog entry. I'm also wondering why it's nullable. Under what circumstances would the descriptor not be available? It'll also need a test (for both the non-null and null cases). And at that point it should probably be a separate PR.

File: pkgs/jnigen/lib/src/generate_bindings.dart [hunk 10]
```diff
@@ -26,6 +27,38 @@ void collectOutputStream(Stream<List<int>> stream, StringBuffer buffer) =>
     stream.transform(const Utf8Decoder()).forEach(buffer.write);
 
 extension JniGenGenerator on JniGenerator {
+  /// Parses the Java input according to this config and dumps all user-visible
+  /// AST nodes.
+  ///
+  /// Returns the formatted dump string.
+  Future<String> dumpApi({Logger? logger}) async {
+    logger ??= createDefaultLogger();
```
Comment (liamappelbe) [inline pkgs/jnigen/lib/src/generate_bindings.dart:35]: Most of this function duplicates line 67-96 of `generate`. Can you factor out the logic into a helper? Looks like the helper could just take a `Logger?` and return `userClasses`.

File: pkgs/jnigen/CHANGELOG.md [hunk 11]
```diff
@@ -30,6 +30,7 @@
 - Allow interface mixin names to be customized using the visitor API.
 - Support customizing the command `java -jar ApiSummarizer.jar` via the
   `summarizer.command` field in config.
+- Add `dumpApi` utility and `ApiDumperVisitor` to dump parsed Java AST.
```
Comment (liamappelbe) [inline pkgs/jnigen/CHANGELOG.md:33]: Move this to the changelog entry for `1.0.1-wip`, and bump it to `1.1.0-wip`. You'll also need to update the jnigen pubspec, and I think there's a const variable somewhere in jnigen that holds the version too.

As of https://github.com/dart-lang/native/pull/3627, you at least won't need to update all the bindings files 😅

**Review by liamappelbe — COMMENTED** [review]
Almost there. Just a few more details.
